docs/advice.org

```diff
@@ -0,0 +1,859 @@
+(import-macros {: defn
+                : defadvice} :lib.advice.macros)
+(local {: remove-advice} :lib.advice)
+
+(defn original-fn
+      [x y z]
+      "docstr"
+      "default")
+
+(original-fn)
+;; => "default"
+
+(defadvice advice-fn
+           [x y z]
+           :override original-fn
+           "over-it")
+;; => "over-it"
+
+(remove-advice original-fn :override advice-fn)
+
+(original-fn)
+;; => "default'
+#+end_src
+
+** Get Advice For an Advisable Function
+
+When testing or debugging it may be useful to see the list of advice
+applied to an advisable function.
+
+The =get-advice= function will do just that:
+
+#+begin_src fennel
+(import-macros {: defn
+                : defadvice} :lib.advice.macros)
+(local {: get-advice} :lib.advice)
+
+(defn original-fn
+      [x y z]
+      "docstr"
+      "default")
+
+(defadvice advice-fn
+           [x y z]
+           :override original-fn
+           "over-it")
+
+(pprint (get-advice original-fn))
+#+end_src
+
+Will print a table like the following:
+
+#+begin_src fennel
+[
+ {:f    "advice-fn: 0x600000278c80"
+  :type "override"}
+]
+#+end_src
+
+** Log Advisable Functions
+
+It may be useful to see a list of advisable function keys. Use the
+=print-advisable-keys= function to print a nicely formatted list of
+advisable keys.
+
+Example:
+
+#+begin_src fennel
+(local {: print-advisable-keys} :lib.advice)
+
+(print-advisable-keys)
+#+end_src
+
+Which would print something like:
+
+#+begin_example
+:test/advice-test/test-func-1
+:test/advice-test/test-func-2
+:test/advice-test/test-func-3
+;; ...
+:test/advice-test/test-func-7
+#+end_example
+
+* Considerations
+
+** Performance
+
+Creating advisable functions does come with some runtime overhead
+iterating through the advice. In most cases the performance hit should
+be negligible,  but if anyone does experience unexpected performance
+issues please report it so maintainers can investigate.
+
+Functions that fire on a short interval, such as animation functions
+that run every 5 milliseconds, may encounter degraded performance
+caused by the advising overhead. It's not recommended to make
+functions like that advisable.
+
+** Complexity
+
+Just like with emacs, use advisable functions cautiously when it's the
+best choice for users to customize behaviors.
+
+** Tables vs. Functions
+
+The =make-advisable= function and =defadvice= macro return tables with a
+=__call=, =__index=, and =__name= metatable entries. The resulting tables
+can be called just like functions, but if you run =(type
+defn-func-2-advice)= it may return  "table" instead of function. If
+this causes any issues, please report it so we can consider alternatives.
+
+* Prior Art
+
+This concept was directly inspired and arguably ripped-off of emacs'
+advising system. Much of their docs are relevant to this, if you would
+like to dig deeper check out the official [[emacs advising
+docs][https://www.gnu.org/software/emacs/manual/html_node/elisp/Advising-Functions.html]]
```

Comment:
No idea why, but this also doesn't render correctly.

Reply:
I think I reversed the order, will try and see